005_hackerrank_challenges/001_introduction/005_filter_array/solution2.hs

```diff
@@ -1,14 +1,19 @@
+#! /usr/bin/env nix-shell
+#! nix-shell ../../../004_an_updated_nix_shell_file/shell.nix  -i runghc
+
 -- Same as solution.hs, but we don't use the builtin filter function.
 -- Instead, as recommended, we define our own custom filter function
+--
+-- run it locally by doing `cat sample-input.txt | ./solution2.hs`
 
 f :: Int -> [Int] -> [Int]
 f n arr = filter' (<n) arr
 
 filter' :: (a -> Bool) -> [a] -> [a]
-filter' f [] = []
-filter' f (x:xs) = if (f x)
-                   then x:filter' f xs
-                   else filter' f xs
+filter' _ [] = []
+filter' g (x:xs) = if g x
+                   then x:filter' g xs
+                   else filter' g xs
```

Comment:
When you aren't using a binding in a pattern make it explicit by using `_` . Also removed unnecessary parentheses and also changed the inner bind `f` to `g` to avoid name shadowing. All suggestions given by hlint